python/src/trezorlib/cli/firmware.py

```diff
@@ -166,39 +281,212 @@ def version_str(version):
+            # Trezor One does not send ButtonRequest
+            click.echo("Please confirm the action on your Trezor device")
+        firmware.update(client, firmware_data)
+    except exceptions.Cancelled:
+        click.echo("Update aborted on device.")
+    except exceptions.TrezorException as e:
+        click.echo(f"Update failed: {e}")
+        sys.exit(3)
+
+
+@click.group(name="firmware")
+def cli():
+    """Firmware commands."""
+
+
+@cli.command()
+# fmt: off
+@click.argument("filename", type=click.File("rb"))
+@click.option("-c", "--check-device", is_flag=True, help="Validate device compatibility")
+@click.option("--fingerprint", help="Expected firmware fingerprint in hex")
+@click.pass_obj
+# fmt: on
+def verify(
+    obj: TrezorConnection,
+    filename: BinaryIO,
+    check_device: bool,
+    fingerprint: str,
+) -> None:
+    """Verify the integrity of the firmware data stored in a file.
+
+    By default the device is not checked and does not need to be connected.
+    Its validation must be specified.
+
+    In case of validation failure exits with the appropriate exit code.
+    """
+    # Deciding if to take the device into account
+    if check_device:
+        with obj.client_context() as client:
+            bootloader_onev2 = _is_bootloader_onev2(client)
+            trezor_major_version = client.features.major_version
+    else:
+        bootloader_onev2 = None
+        trezor_major_version = None
+
+    firmware_data = filename.read()
+    validate_firmware(
+        firmware_data=firmware_data,
+        fingerprint=fingerprint,
+        bootloader_onev2=bootloader_onev2,
+        trezor_major_version=trezor_major_version,
+    )
+
+
+@cli.command()
+# fmt: off
+@click.option("-o", "--output", type=click.File("wb"), help="Output file to save firmware data to")
+@click.option("-v", "--version", help="Which version to download")
+@click.option("-s", "--skip-check", is_flag=True, help="Do not validate firmware integrity")
+@click.option("--beta", is_flag=True, help="Use firmware from BETA channel")
+@click.option("--bitcoin-only", is_flag=True, help="Use bitcoin-only firmware (if possible)")
+@click.option("--fingerprint", help="Expected firmware fingerprint in hex")
+# fmt: on
+@with_client
+def download(
+    client: TrezorClient,
+    output: BinaryIO,
+    version: str,
+    skip_check: bool,
+    fingerprint: str,
+    beta: bool,
+    bitcoin_only: bool,
+) -> None:
+    """Download and save the firmware image.
+
+    Validation is done by default, can be omitted by "-s" or "--skip-check".
+    """
+    url, fp = find_best_firmware_version(
+        client=client, version=version, beta=beta, bitcoin_only=bitcoin_only
+    )
+
+    firmware_data = download_firmware_data(url)
 
-    return url, fingerprint
+    # When fingerprint and output file were not supplied, take them from SL servers
+    if not fingerprint:
+        fingerprint = fp
+    if not output:
+        output = open(_get_file_name_from_url(url), "wb")
```

Comment:
please also move this after the validation. otherwise you'd create a new file and leave it empty if validation failed

Reply:
Fixed in 015edd3